### DISCUSSION TITLE
About database support in fresh

### DISCUSSION THREAD

**Discussion by LoganTann** [freshframework/fresh#436] [General]
Hello,

After a talk in issue #336, we move here to express our opinion about the idea to integrate a database support directly in the fresh framework.

Here is the context : 

> I really want to try out this new framework.
> 
> But I notice that the docs don't say anything about databases.
> 
> I mean Symfony has Doctrine, Laravel Eloquent.
> 
> I think some kind of ORM is something that's just part of a full-stack framework.
> Don't you think so?
> 
> Or am I missing something? How do you deal with databases?

**Comment by michalvavra** [comment]
I was experimenting with setting up [kysely](https://github.com/koskimas/kysely) query builder. I'm not fan of ORMs.
Setup was very straightforward and I will look more into it in the future.
Very simple example repo: https://github.com/michalvavra/fresh-sqlite-example

Personally I would prefer database/ORM setup as part of examples, instead of built-in functionality.
What I would really like to see is support for [oak](https://github.com/oakserver/oak) middlewares.

**Reply by alexgleason** [reply]
This is the correct answer to the question I wasn't sure how to ask.

The question is along the lines of: does Fresh process React files at runtime as part of its main event loop? The answer is seemingly "yes", allowing you to do the intuitive thing of just creating new TypeScript files and importing them where you need them.

Experience with tools like Next.js poisons your brain and actually makes it **harder to understand new things as an experienced person** than as a novice.

**Reply by marvinhagemeister** [reply]
> The question is along the lines of: does Fresh process React files at runtime as part of its main event loop? The answer is seemingly "yes", allowing you to do the intuitive thing of just creating new TypeScript files and importing them where you need them.

I've read this section a couple of times now, but I have no clue what it means or what it wants to say. Can you rephrase that a little more clearly? I'd like to understand your point of view better as there seems to be something that is bugging you about Fresh. I haven't figured out what that is yet.

**Reply by alexgleason** [reply]
Sure, the question is "if I import my database into 3 different pages, will it create 3 separate database connections no matter what?"

Fresh seems to be a regular `Deno.serve` with a TSX middleware, which is what I want. It's not serving each of my routes in a separate process.

But seeing TSX used anywhere has poisoned my brain into thinking some build process must be involved, leading me to ask unnecessary questions like "is it even possible to use a database?"

It's a dig at the existing JS ecosystem. Fresh is doing a good job here.

**Comment by vorcigernix** [comment]
A relevant question I hope. In the example repo there is an API example and in docs there is an example with a page rendered on both server and client side and passing data between them (similar approach to getserversideprops). What would you do for an admin-like app that needs to load data from the database and insert new records? 
My current approach is to write an API and call it from the server side for displaying/islands for inserting. Alternatively, I can write the reading into the page handler and updates into the island components. Thoughts? My db is postgre on supabase.

**Reply by vorcigernix** [reply]
Worth mentioning: I was thinking about API primarily because in future, data loads would be probably a good fit for some caching scenario.

**Reply by michalvavra** [reply]
`api/` sounds good.

Some sort of CSRF protection would be good (to mitigate random bot attacks).
I was thinking to fork and re-use either [sessions](https://deno.land/x/sessions) or [oak_sessions](https://deno.land/x/oak_sessions) for that.

**Comment by CAYdenberg** [comment]
I've been using CouchDB with Fresh for my [company website](https://livingpixel.io). The advantage is that Couch's interface is pure http, so you don't need a Deno specific driver nor does it matter how many app instances are spun up. I have models defined in zod, configurable rest API, and it can store files (via CouchDB attachments).

Probably needs some cleanup, but if anyone is interested in our process reach out and we can talk some time.
